packages/edit-site/src/components/global-styles/hooks.js

```diff
@@ -0,0 +1,347 @@
+			set( newSettings, path, ( presetData ?? {} ).user );
+		}
+	} );
+	return newSettings;
+}
+
+function useGlobalStylesUserConfig() {
+	const { globalStylesId, content } = useSelect( ( select ) => {
+		const _globalStylesId = select( editSiteStore ).getSettings()
+			.__experimentalGlobalStylesUserEntityId;
+		return {
+			globalStylesId: _globalStylesId,
+			content: select( coreStore ).getEditedEntityRecord(
+				'postType',
+				'wp_global_styles',
+				_globalStylesId
+			)?.content,
+		};
+	}, [] );
+	const { getEditedEntityRecord } = useSelect( coreStore );
+	const { editEntityRecord } = useDispatch( coreStore );
+
+	const parseContent = ( contentToParse ) => {
+		let parsedConfig;
+		try {
+			parsedConfig = contentToParse ? JSON.parse( contentToParse ) : {};
+			// It is very important to verify if the flag isGlobalStylesUserThemeJSON is true.
+			// If it is not true the content was not escaped and is not safe.
+			if ( ! parsedConfig.isGlobalStylesUserThemeJSON ) {
+				parsedConfig = {};
+			} else {
+				parsedConfig = {
+					...parsedConfig,
+					settings: addUserOriginToSettings( parsedConfig.settings ),
+				};
+			}
+		} catch ( e ) {
+			/* eslint-disable no-console */
+			console.error( 'Global Styles User data is not valid' );
+			console.error( e );
+			/* eslint-enable no-console */
+			parsedConfig = {};
+		}
+
+		return parsedConfig;
+	};
+
+	const config = useMemo( () => {
+		return parseContent( content );
+	}, [ content ] );
+
+	const setConfig = useCallback(
+		( callback ) => {
+			const currentConfig = parseContent(
+				getEditedEntityRecord(
+					'postType',
+					'wp_global_styles',
+					globalStylesId
+				)?.content
+			);
+			const updatedConfig = callback( currentConfig );
+			editEntityRecord( 'postType', 'wp_global_styles', globalStylesId, {
+				content: JSON.stringify( {
+					...updatedConfig,
+					settings: removeUserOriginFromSettings(
+						updatedConfig.settings
+					),
+				} ),
+			} );
+		},
+		[ globalStylesId ]
+	);
+
+	return [ config, setConfig ];
+}
+
+function useGlobalStylesBaseConfig() {
+	const baseConfig = useSelect( ( select ) => {
+		return select( editSiteStore ).getSettings()
+			.__experimentalGlobalStylesBaseStyles;
+	}, [] );
+
+	return baseConfig;
+}
+
+export function useGlobalStylesConfig() {
+	const [ userConfig, setUserConfig ] = useGlobalStylesUserConfig();
+	const baseConfig = useGlobalStylesBaseConfig();
+	const mergedConfig = useMemo( () => {
+		return mergeBaseAndUserConfigs( baseConfig, userConfig );
+	}, [ userConfig, baseConfig ] );
+
+	return [ baseConfig, userConfig, mergedConfig, setUserConfig ];
+}
+
+export const useGlobalStylesReset = () => {
+	const [ config, setConfig ] = useGlobalStylesUserConfig();
+	const canReset = !! config && ! isEqual( config, EMPTY_CONFIG );
+	return [
+		canReset,
+		useCallback( () => setConfig( () => EMPTY_CONFIG ), [ setConfig ] ),
+	];
+};
+
+export function useSetting( path, blockName, source = 'all' ) {
```

Comment:
The useSetting hook on the block editor has a slightly different API the source is part of the path.
useSetting( 'color.palette' ) returns merged color palette.
useSetting( 'color.palette.user' ) returns user color palette.

We may need to consolidate these two approaches.